Firestore/core/src/api/document_snapshot.cc

```diff
@@ -88,7 +88,12 @@ absl::optional<google_firestore_v1_Value> DocumentSnapshot::GetValue(
 bool operator==(const DocumentSnapshot& lhs, const DocumentSnapshot& rhs) {
   return lhs.firestore_ == rhs.firestore_ &&
          lhs.internal_key_ == rhs.internal_key_ &&
-         lhs.internal_document_ == rhs.internal_document_ &&
+         lhs.internal_document_.has_value() ==
+             rhs.internal_document_.has_value() &&
+         (lhs.internal_document_.has_value()
+              ? lhs.internal_document_->get().CompareIgnoringState(
```

Comment:
Nit (or personal preference): What do you think of comparing the value and the key from the document, and then using `exists()` and `snapshotMetadata` for the comparison of the remaining state (all encapsulated in this equals function).

Reply:
Good idea. Done.